### PR TITLE
Runtime State Test + Integration with `try-runtime`

### PR DESCRIPTION
**Pull request by kianenigma**
This PR adds a new hook to the `try-runtime follow-chain` subcommand to execute a custom hook called `sanity_check` per block. This allows us to avoid dumping too much testing code into `debug_assertion`, and instead test them in real life. To see more about what `follow-chain` itself is doing, see #9788. 

Polkadot companion: https://github.com/paritytech/polkadot/pull/5907

## (Updated) Context: 

I have pulled this PR out of the grave again, since I need it for my nomination pools. Reviews can be easy going, since this is all offchain code, but, please be aware: 

1. I genuinely dislike the name `sanity_check` now. I think we should rename it to `on_block_check`, and rename `integrity_test`to `on_construct_runtime_check`. If all agree, I will do it in this PR.
2. I had to add a new type `AllPalletWithSystemFlat`, which is basically the same thing as before but tuples are not nested. This is needed because `PalletInfoAccess` is implemented on each individual pallet, but not on tuples. In reviews, please double check for me that this has no side-effect, other than us needing to increase some of the limits on `impl_trait_for_tuples`  from 30 to 100.   

Currently, the `follow-chain` command is an okay-ish way to continuously run these sanity checks, but honestly it is kinda slow and sluggish. As a follow-up, I will work on integrating this into the main client, so you can import blocks but use the try-runtime api instead of Core for block import. 

#### TODO:

1. [x] think of a wrapper trait to hide the second parameter.
2. [x] feature gate functions again.
4. [x] fix all tests.
5. [x] double check that `AllPalletsWithSystemFlat` has not side-effect.
6. [ ] Clean up the return type of sanity checks
7. [ ] (optional) Rename `integrity_check` -> `on_construct_runtime_check`
8. [x] (optional) Rename `sanity_check` -> `state_consistency_check`

### CODE REVIEW

**Comment by kianenigma** [comment]
yes there is.

**Comment by kianenigma** [comment]
stale but I will get to it soon.

**Comment by kianenigma** [comment]
want this.

**Comment by kianenigma** [comment]
YES.

File: frame/executive/src/lib.rs
Comment (kianenigma) [reply frame/executive/src/lib.rs:229]: They are already entirely `#cfg[(try-runtime)]`.

File: frame/executive/src/lib.rs
Comment (kianenigma) [reply frame/executive/src/lib.rs:303]: `custom`

File: frame/executive/src/lib.rs
Comment (kianenigma) [reply frame/executive/src/lib.rs:230]: that's a good idea, will rename.

File: frame/support/src/traits/hooks.rs
Comment (xlc) [inline frame/support/src/traits/hooks.rs:33]: make number of pallets in a runtime is 256 so we might as well make this 256

File: frame/executive/src/lib.rs
Comment (ggwpez) [inline frame/executive/src/lib.rs:253]: Looks like to me that you are doing nearly all checks from `final_checks`, but copy paste.  
The State root is also checked?!

File: frame/staking/src/pallet/impls.rs
Comment (ggwpez) [inline frame/staking/src/pallet/impls.rs:1457]: I think the renaming to `integrity_check` is still a good idea.

File: frame/support/src/traits/hooks.rs
Comment (ggwpez) [inline frame/support/src/traits/hooks.rs:136]:
```suggestion
			SanityCheckTargets::RoundRobin(x) => write!(f, "RoundRobin({})", x),
```

**Review by ggwpez — APPROVED** [review]

**Comment by kianenigma** [comment]
This is blocked by https://github.com/paritytech/substrate/pull/11813

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1743788

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1749329

File: frame/support/procedural/src/construct_runtime/mod.rs
Comment (kianenigma) [reply frame/support/procedural/src/construct_runtime/mod.rs:348]: Outdated.

File: frame/staking/src/pallet/impls.rs
Comment (kianenigma) [reply frame/staking/src/pallet/impls.rs:1458]: But then this name is taken for something else :( 

I will probably rename them in this PR as well.

File: frame/executive/src/lib.rs
Comment (kianenigma) [reply frame/executive/src/lib.rs:253]: the state root is checked IFF the `check_state_root: bool` is set. Updated the doc to reflect this.

File: frame/support/src/traits/hooks.rs
Comment (kianenigma) [reply frame/support/src/traits/hooks.rs]: Settled in #11813

File: frame/executive/src/lib.rs
Comment (kianenigma) [reply frame/executive/src/lib.rs:230]: fixed.

File: bin/node-template/runtime/src/lib.rs
Comment (kianenigma) [inline bin/node-template/runtime/src/lib.rs:553]:
```suggestion
			Executive::try_execute_block(block, state_root_check, select).expect("execute-block failed")
```

File: frame/support/src/traits/try_runtime.rs
Comment (muharem) [reply frame/support/src/traits/try_runtime.rs:149]: ah, for live mode

**Review by muharem — APPROVED** [review]
my humble approve for the topic where I am not an expert.

File: frame/support/src/traits/try_runtime.rs
Comment (dvdplm) [inline frame/support/src/traits/try_runtime.rs:34]: This is automatically cleaned up right? Should the name and docs reflect that, `temp_storage_key`?

File: frame/support/src/traits/try_runtime.rs
Comment (dvdplm) [inline frame/support/src/traits/try_runtime.rs:147]: Maybe easier to read this way?
```suggestion
let skip: u32 = n.clone().try_into().ok().expect("Block numbers can be cast to a u32; qed");
let skip = skip % len;
```

File: frame/support/src/traits/try_runtime.rs
Comment (dvdplm) [inline frame/support/src/traits/try_runtime.rs:164]: The `.as_bytes().to_vec()` is a bit annoying here. I imagine `for_tuples!` need to return `&'static str` for some reason? Maybe we can `map` above when assigning to `functions`?

File: utils/frame/try-runtime/cli/src/commands/execute_block.rs
Comment (dvdplm) [inline utils/frame/try-runtime/cli/src/commands/execute_block.rs:31]:
```suggestion
/// This will always call into `TryRuntime_execute_block`, which can optionally skip the state-root
```

Comment (dvdplm) [reply utils/frame/try-runtime/cli/src/commands/execute_block.rs:31]: Did you mean to write `TryRuntime::execute_block` here?

File: utils/frame/try-runtime/cli/src/commands/execute_block.rs
Comment (dvdplm) [inline utils/frame/try-runtime/cli/src/commands/execute_block.rs:39]:
```suggestion
	/// If set the state root check is disabled.
```

File: utils/frame/try-runtime/cli/src/lib.rs
Comment (dvdplm) [inline utils/frame/try-runtime/cli/src/lib.rs:346]: Maybe we can provide doc-links to these?

File: frame/executive/src/lib.rs
Comment (bkchr) [inline frame/executive/src/lib.rs:262]: Assuming you are running with a different runtime, this could be easily something that is different :see_no_evil: 

Maybe we should just disable all checks if wanted.

File: frame/support/procedural/src/pallet/expand/hooks.rs
Comment (bkchr) [inline frame/support/procedural/src/pallet/expand/hooks.rs:66]: You can use an expect here. We use it everywhere.

```
expect("Every active pallet has a name in the runtime; qed")
```

File: frame/support/src/traits/try_runtime.rs
Comment (bkchr) [inline frame/support/src/traits/try_runtime.rs:113]: Users wanting to use the CLI will have 0 clue on what kind of values are accepted. This needs to be documented in the CLI.

File: frame/support/src/traits/try_runtime.rs
Comment (bkchr) [inline frame/support/src/traits/try_runtime.rs:164]:
```suggestion
					if pallet_names.iter().any(|n| n == name.as_bytes()) {
```

**Review by bkchr — APPROVED** [review]

File: frame/support/src/traits/try_runtime.rs
Comment (kianenigma) [reply frame/support/src/traits/try_runtime.rs:149]: This is the `RoundRobin` one where we want to run e.g. 5 checks per block, so we need to skip some.

File: frame/support/src/traits/try_runtime.rs
Comment (kianenigma) [reply frame/support/src/traits/try_runtime.rs:147]: Thanks!

File: frame/support/src/traits/try_runtime.rs
Comment (kianenigma) [reply frame/support/src/traits/try_runtime.rs:34]: This key is only ever used in pre/post migrate hooks. Even if we store some data there and not remove it, it is okay.

File: frame/support/src/traits/try_runtime.rs
Comment (kianenigma) [reply frame/support/src/traits/try_runtime.rs:164]: To check a `&'static str` (which is what PalletInfo.name() gives you) against a `Vec<u8>`, at some point you have to do `str.as_bytes().to_vec()`. I don't see much difference between doing it here or in `for_tuples!`.

File: utils/frame/try-runtime/cli/src/commands/execute_block.rs
Comment (kianenigma) [reply utils/frame/try-runtime/cli/src/commands/execute_block.rs:31]: Nope, it is intentional. The method that's called is actually `TryRuntime_execute_block` based on the rules of how runtime APIs are generated.

File: utils/frame/try-runtime/cli/src/lib.rs
Comment (kianenigma) [reply utils/frame/try-runtime/cli/src/lib.rs:346]: As noted, these are string literals that are used to call into runtime APIs, nothing to link to: https://github.com/paritytech/substrate/blob/42450b524c7f69d0bba7562aee4d8e9c4ee8f714/utils/frame/try-runtime/cli/src/commands/execute_block.rs#L197

File: frame/bags-list/src/list/mod.rs
Comment (kianenigma) [reply frame/bags-list/src/list/mod.rs:224]: Not automatically yet, I think it is okay to leave it up to the pallet to do so, which is what you are seeing here.

We can make a follow-up and once `try_state` is a bit more standard, make it execute upon all migrations.

File: frame/support/src/traits/try_runtime.rs
Comment (kianenigma) [reply frame/support/src/traits/try_runtime.rs:34]: We don't always implement `OnRuntimeUpgrade` and this for `Pallet<T>`. In most cases, they are implemented for standalone structs, so this won't work.

File: frame/executive/src/lib.rs
Comment (kianenigma) [reply frame/executive/src/lib.rs:262]: I will leave that up to https://github.com/paritytech/substrate/pull/12048#issuecomment-1228427709 and @pmikolajczyk41

File: frame/support/procedural/src/pallet/expand/hooks.rs
Comment (kianenigma) [reply frame/support/procedural/src/pallet/expand/hooks.rs:66]: sure.

File: frame/support/src/traits/try_runtime.rs
Comment (kianenigma) [reply frame/support/src/traits/try_runtime.rs:113]: great point!

File: frame/support/src/traits/try_runtime.rs
Comment (kianenigma) [reply frame/support/src/traits/try_runtime.rs:164]: well of course this man has a better solution :p https://github.com/paritytech/substrate/pull/10174#discussion_r955956139

**Comment by kianenigma** [comment]
@bkchr @ggwpez @muharem can you review the companion as well?

File: frame/support/src/traits/try_runtime.rs
Comment (dvdplm) [inline frame/support/src/traits/try_runtime.rs:122]: Stray typo?

File: frame/support/src/traits/try_runtime.rs
Comment (dvdplm) [inline frame/support/src/traits/try_runtime.rs:125]: I don't understand what the part about "to be help" means. Rephrase?

**Review by dvdplm — APPROVED** [review]
Two doc nitpicks, otherwise lgtm.

File: frame/support/src/traits/try_runtime.rs
Comment (bkchr) [reply frame/support/src/traits/try_runtime.rs:34]: Ahh yeah, I'm an idiot :facepalm:

File: frame/support/src/traits/try_runtime.rs
Comment (kianenigma) [reply frame/support/src/traits/try_runtime.rs:125]: I meant "held" 🤦 sorry

**Comment by kianenigma** [comment]
This will be merged, but it will be very buggy until https://github.com/paritytech/substrate/pull/12048 is merged.

**Comment by kianenigma** [comment]
bot merge

**Comment by ggwpez** [comment]
bot rebase

**Comment by kianenigma** [comment]
bot merge

**Comment by shawntabrizi** [comment]
NOTE FOR DEVELOPERS:

This PR triggers a lot of errors of the `try-runtime` feature flag is not correctly set up on your runtime.

Make sure each custom pallet has at least a `try-runtime` feature flag:

```
try-runtime = [ "frame-support/try-runtime" ]
```

and that all of these features are enabled across the pallets used in your runtime:

See: https://github.com/paritytech/cumulus/pull/1551/commits/480b977ac7c9a7e94c0ce3b4599d71c22546ab84


Otherwise you will get errors like:

```
error: failed to run custom build command for `statemint-runtime v1.0.0 (/Users/shawntabrizi/Documents/GitHub/cumulus/parachains/runtimes/assets/statemint)`

Caused by:
  process didn't exit successfully: `/Users/shawntabrizi/Documents/GitHub/cumulus/target/debug/build/statemint-runtime-7ab48e0f89a28cba/build-script-build` (exit status: 1)
  --- stdout
  Information that should be included in a bug report.
  Executing build command: "/Users/shawntabrizi/.rustup/toolchains/nightly-aarch64-apple-darwin/bin/cargo" "rustc" "--target=wasm32-unknown-unknown" "--manifest-path=/Users/shawntabrizi/Documents/GitHub/cumulus/target/debug/wbuild/statemint-runtime/Cargo.toml" "--color=always" "--profile" "release"
  Using rustc version: rustc 1.65.0-nightly (34a6cae28 2022-08-09)


  --- stderr
     Compiling pallet-utility v4.0.0-dev (https://github.com/paritytech/substrate?branch=master#5b0f999d)
     Compiling statemint-runtime v1.0.0 (/Users/shawntabrizi/Documents/GitHub/cumulus/parachains/runtimes/assets/statemint)
  error[E0599]: the function or associated item `try_runtime_upgrade` exists for struct `frame_executive::Executive<Runtime, sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, sp_runtime::generic::UncheckedExtrinsic<MultiAddress<sp_runtime::AccountId32, ()>, Call, MultiSignature, (CheckNonZeroSender<Runtime>, CheckSpecVersion<Runtime>, CheckTxVersion<Runtime>, CheckGenesis<Runtime>, CheckEra<Runtime>, CheckNonce<Runtime>, CheckWeight<Runtime>, ChargeAssetTxPayment<Runtime>)>>, ChainContext<Runtime>, Runtime, (frame_system::Pallet<Runtime>, cumulus_pallet_parachain_system::Pallet<Runtime>, pallet_timestamp::Pallet<Runtime>, parachain_info::Pallet<Runtime>, pallet_balances::Pallet<Runtime>, pallet_transaction_payment::Pallet<Runtime>, pallet_asset_tx_payment::Pallet<Runtime>, pallet_authorship::Pallet<Runtime>, pallet_collator_selection::Pallet<Runtime>, pallet_session::Pallet<Runtime>, pallet_aura::Pallet<Runtime>, cumulus_pallet_aura_ext::Pallet<Runtime>, cumulus_pallet_xcmp_queue::Pallet<Runtime>, pallet_xcm::Pallet<Runtime>, cumulus_pallet_xcm::Pallet<Runtime>, cumulus_pallet_dmp_queue::Pallet<Runtime>, pallet_utility::Pallet<Runtime>, pallet_multisig::Pallet<Runtime>, pallet_proxy::Pallet<Runtime>, pallet_assets::Pallet<Runtime>, pallet_uniques::Pallet<Runtime>)>`, but its trait bounds were not satisfied
     --> /Users/shawntabrizi/Documents/GitHub/cumulus/parachains/runtimes/assets/statemint/src/lib.rs:805:28
      |
  805 |             let weight = Executive::try_runtime_upgrade().unwrap();
      |                                     ^^^^^^^^^^^^^^^^^^^ function or associated item cannot be called on `frame_executive::Executive<Runtime, sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, sp_runtime::generic::UncheckedExtrinsic<MultiAddress<sp_runtime::AccountId32, ()>, Call, MultiSignature, (CheckNonZeroSender<Runtime>, CheckSpecVersion<Runtime>, CheckTxVersion<Runtime>, CheckGenesis<Runtime>, CheckEra<Runtime>, CheckNonce<Runtime>, CheckWeight<Runtime>, ChargeAssetTxPayment<Runtime>)>>, ChainContext<Runtime>, Runtime, (frame_system::Pallet<Runtime>, cumulus_pallet_parachain_system::Pallet<Runtime>, pallet_timestamp::Pallet<Runtime>, parachain_info::Pallet<Runtime>, pallet_balances::Pallet<Runtime>, pallet_transaction_payment::Pallet<Runtime>, pallet_asset_tx_payment::Pallet<Runtime>, pallet_authorship::Pallet<Runtime>, pallet_collator_selection::Pallet<Runtime>, pallet_session::Pallet<Runtime>, pallet_aura::Pallet<Runtime>, cumulus_pallet_aura_ext::Pallet<Runtime>, cumulus_pallet_xcmp_queue::Pallet<Runtime>, pallet_xcm::Pallet<Runtime>, cumulus_pallet_xcm::Pallet<Runtime>, cumulus_pallet_dmp_queue::Pallet<Runtime>, pallet_utility::Pallet<Runtime>, pallet_multisig::Pallet<Runtime>, pallet_proxy::Pallet<Runtime>, pallet_assets::Pallet<Runtime>, pallet_uniques::Pallet<Runtime>)>` due to unsatisfied trait bounds
      |
      = note: the following trait bounds were not satisfied:
              `(frame_system::Pallet<Runtime>, cumulus_pallet_parachain_system::Pallet<Runtime>, pallet_timestamp::Pallet<Runtime>, parachain_info::Pallet<Runtime>, pallet_balances::Pallet<Runtime>, pallet_transaction_payment::Pallet<Runtime>, pallet_asset_tx_payment::Pallet<Runtime>, pallet_authorship::Pallet<Runtime>, pallet_collator_selection::Pallet<Runtime>, pallet_session::Pallet<Runtime>, pallet_aura::Pallet<Runtime>, cumulus_pallet_aura_ext::Pallet<Runtime>, cumulus_pallet_xcmp_queue::Pallet<Runtime>, pallet_xcm::Pallet<Runtime>, cumulus_pallet_xcm::Pallet<Runtime>, cumulus_pallet_dmp_queue::Pallet<Runtime>, pallet_utility::Pallet<Runtime>, pallet_multisig::Pallet<Runtime>, pallet_proxy::Pallet<Runtime>, pallet_assets::Pallet<Runtime>, pallet_uniques::Pallet<Runtime>): TryState<u32>`

  error[E0599]: the function or associated item `try_execute_block` exists for struct `frame_executive::Executive<Runtime, sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, sp_runtime::generic::UncheckedExtrinsic<MultiAddress<sp_runtime::AccountId32, ()>, Call, MultiSignature, (CheckNonZeroSender<Runtime>, CheckSpecVersion<Runtime>, CheckTxVersion<Runtime>, CheckGenesis<Runtime>, CheckEra<Runtime>, CheckNonce<Runtime>, CheckWeight<Runtime>, ChargeAssetTxPayment<Runtime>)>>, ChainContext<Runtime>, Runtime, (frame_system::Pallet<Runtime>, cumulus_pallet_parachain_system::Pallet<Runtime>, pallet_timestamp::Pallet<Runtime>, parachain_info::Pallet<Runtime>, pallet_balances::Pallet<Runtime>, pallet_transaction_payment::Pallet<Runtime>, pallet_asset_tx_payment::Pallet<Runtime>, pallet_authorship::Pallet<Runtime>, pallet_collator_selection::Pallet<Runtime>, pallet_session::Pallet<Runtime>, pallet_aura::Pallet<Runtime>, cumulus_pallet_aura_ext::Pallet<Runtime>, cumulus_pallet_xcmp_queue::Pallet<Runtime>, pallet_xcm::Pallet<Runtime>, cumulus_pallet_xcm::Pallet<Runtime>, cumulus_pallet_dmp_queue::Pallet<Runtime>, pallet_utility::Pallet<Runtime>, pallet_multisig::Pallet<Runtime>, pallet_proxy::Pallet<Runtime>, pallet_assets::Pallet<Runtime>, pallet_uniques::Pallet<Runtime>)>`, but its trait bounds were not satisfied
     --> /Users/shawntabrizi/Documents/GitHub/cumulus/parachains/runtimes/assets/statemint/src/lib.rs:817:15
      |
  817 |             Executive::try_execute_block(block, state_root_check, select).expect("try_execute_block failed")
      |                        ^^^^^^^^^^^^^^^^^ function or associated item cannot be called on `frame_executive::Executive<Runtime, sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, sp_runtime::generic::UncheckedExtrinsic<MultiAddress<sp_runtime::AccountId32, ()>, Call, MultiSignature, (CheckNonZeroSender<Runtime>, CheckSpecVersion<Runtime>, CheckTxVersion<Runtime>, CheckGenesis<Runtime>, CheckEra<Runtime>, CheckNonce<Runtime>, CheckWeight<Runtime>, ChargeAssetTxPayment<Runtime>)>>, ChainContext<Runtime>, Runtime, (frame_system::Pallet<Runtime>, cumulus_pallet_parachain_system::Pallet<Runtime>, pallet_timestamp::Pallet<Runtime>, parachain_info::Pallet<Runtime>, pallet_balances::Pallet<Runtime>, pallet_transaction_payment::Pallet<Runtime>, pallet_asset_tx_payment::Pallet<Runtime>, pallet_authorship::Pallet<Runtime>, pallet_collator_selection::Pallet<Runtime>, pallet_session::Pallet<Runtime>, pallet_aura::Pallet<Runtime>, cumulus_pallet_aura_ext::Pallet<Runtime>, cumulus_pallet_xcmp_queue::Pallet<Runtime>, pallet_xcm::Pallet<Runtime>, cumulus_pallet_xcm::Pallet<Runtime>, cumulus_pallet_dmp_queue::Pallet<Runtime>, pallet_utility::Pallet<Runtime>, pallet_multisig::Pallet<Runtime>, pallet_proxy::Pallet<Runtime>, pallet_assets::Pallet<Runtime>, pallet_uniques::Pallet<Runtime>)>` due to unsatisfied trait bounds
      |
      = note: the following trait bounds were not satisfied:
              `(frame_system::Pallet<Runtime>, cumulus_pallet_parachain_system::Pallet<Runtime>, pallet_timestamp::Pallet<Runtime>, parachain_info::Pallet<Runtime>, pallet_balances::Pallet<Runtime>, pallet_transaction_payment::Pallet<Runtime>, pallet_asset_tx_payment::Pallet<Runtime>, pallet_authorship::Pallet<Runtime>, pallet_collator_selection::Pallet<Runtime>, pallet_session::Pallet<Runtime>, pallet_aura::Pallet<Runtime>, cumulus_pallet_aura_ext::Pallet<Runtime>, cumulus_pallet_xcmp_queue::Pallet<Runtime>, pallet_xcm::Pallet<Runtime>, cumulus_pallet_xcm::Pallet<Runtime>, cumulus_pallet_dmp_queue::Pallet<Runtime>, pallet_utility::Pallet<Runtime>, pallet_multisig::Pallet<Runtime>, pallet_proxy::Pallet<Runtime>, pallet_assets::Pallet<Runtime>, pallet_uniques::Pallet<Runtime>): TryState<u32>`

  For more information about this error, try `rustc --explain E0599`.
  error: could not compile `statemint-runtime` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
```

**Pull request merged**
